File: frontend/src/components/DH/DiffieHellman.js
/* global BigInt */
class DiffieHellman {
    constructor(bitLength = 128) {
        this.bitLength = bitLength;
        this.p = this.generateProbablePrime(bitLength);
        this.g = this.findPrimitiveRootSimple(this.p);
        this.privateKey = this.generatePrivateKey();
        this.publicKey = this.generatePublicKey();
    }

    generateRandomBigInt(bitLength) {
        const byteLength = Math.ceil(bitLength / 8);
        const randomBytes = new Uint8Array(byteLength);
        crypto.getRandomValues(randomBytes);

        randomBytes[0] |= 0x80;
        randomBytes[byteLength - 1] |= 0x01;

        let hexString = '0x';
        randomBytes.forEach(byte => {
            hexString += byte.toString(16).padStart(2, '0');
        });

        return BigInt(hexString);
    }

    generateProbablePrime(bitLength, certainty = 20) {
        let candidate;
        do {
            candidate = this.generateRandomBigInt(bitLength);
        } while (!this.isProbablePrime(candidate, certainty));

        return candidate;
    }

    isProbablePrime(n, k) {
        if (n <= 1n) return false;
        if (n <= 3n) return true;
        if (n % 2n === 0n) return false;

        let d = n - 1n;
        let s = 0n;
        while (d % 2n === 0n) {
            d /= 2n;
            s++;
        }

        for (let i = 0; i < k; i++) {
            const a = this.getRandomBigInt(2n, n - 2n);
            let x = this.modPow(a, d, n);

            if (x === 1n || x === n - 1n) continue;

            let j;
            for (j = 1n; j < s; j++) {
                x = this.modPow(x, 2n, n);
                if (x === n - 1n) break;
            }

            if (j === s) return false;
        }

        return true;
    }

    findPrimitiveRootSimple(p) {
        if (p === 2n) return 1n;
        const pMinus1 = p - 1n;

        const candidates = [2n, 3n, 5n, 7n, 11n];
        for (const g of candidates) {
            if (this.modPow(g, pMinus1, p) === 1n) {
                return g;
            }
        }

        throw new Error('Failed to find primitive root');
    }

    generatePrivateKey() {
        const min = 2n;
        const max = this.p - 2n;
        const range = max - min;
        const byteLength = Math.ceil(this.bitLength / 8);
        const randomBytes = new Uint8Array(byteLength);

        let privateKey;
        do {
            crypto.getRandomValues(randomBytes);
            let hexString = '0x';
            randomBytes.forEach(byte => {
                hexString += byte.toString(16).padStart(2, '0');
            });
            privateKey = BigInt(hexString) % range + min;
        } while (privateKey <= 1n || privateKey >= this.p - 1n);

        return privateKey;
    }

    generatePublicKey() {
        return this.modPow(this.g, this.privateKey, this.p);
    }

    computeSharedSecret(otherPublicKey) {
        return this.modPow(otherPublicKey, this.privateKey, this.p);
    }

    modPow(base, exponent, modulus) {
        if (modulus === 1n) return 0n;
        let result = 1n;
        base = base % modulus;

        while (exponent > 0n) {
            if (exponent % 2n === 1n) {
                result = (result * base) % modulus;
            }
            exponent = exponent >> 1n;
            base = (base * base) % modulus;
        }

        return result;
    }

    getRandomBigInt(min, max) {
        const range = max - min;
        const byteLength = Math.ceil(range.toString(2).length / 8);
        const randomBytes = new Uint8Array(byteLength);

        let result;
        do {
            crypto.getRandomValues(randomBytes);
            let hexString = '0x';
            randomBytes.forEach(byte => {
                hexString += byte.toString(16).padStart(2, '0');
            });
            result = BigInt(hexString) % range + min;
        } while (result < min || result > max);

        return result;
    }
}

export default DiffieHellman;

File: Cryptography/DiffieHellman/DiffieHellman.cs
using System.Numerics;
using System.Security.Cryptography;

namespace Cryptography.DiffieHellman;

public class DiffieHellman
{
        private static readonly RandomNumberGenerator rng = RandomNumberGenerator.Create();
    
    public static (BigInteger g, BigInteger p) GenerateParameters(int bitLength)
    {
        BigInteger p = GenerateProbablePrime(bitLength);
        BigInteger g = FindPrimitiveRootSimple(p);
        return (g, p);
    }

    private static BigInteger GenerateProbablePrime(int bitLength)
    {
        byte[] bytes = new byte[bitLength / 8 + 1];
        BigInteger p;
    
        do {
            rng.GetBytes(bytes);
            bytes[^1] &= (byte)(0xFF >> (8 - bitLength % 8));
            p = new BigInteger(bytes, isUnsigned: true);
            p |= BigInteger.One;
        } while (!IsProbablePrime(p, 20));
    
        return p;
    }

    private static BigInteger FindPrimitiveRootSimple(BigInteger p)
    {
        if (p == 2) return 1;
        BigInteger pMinus1 = p - 1;
    
        foreach (var g in new[] { 2, 3, 5, 7, 11 })
        {
            if (BigInteger.ModPow(g, pMinus1, p) == 1)
                return g;
        }
    
        throw new Exception("Failed to find primitive root");
    }

    private static bool IsProbablePrime(BigInteger n, int certainty)
    {
        if (n < 2) return false;
        if (n == 2 || n == 3) return true;
        if (n % 2 == 0) return false;
        
        BigInteger d = n - 1;
        int s = 0;
        
        while (d % 2 == 0)
        {
            d /= 2;
            s++;
        }

        byte[] bytes = new byte[n.ToByteArray().LongLength];

        for (int i = 0; i < certainty; i++)
        {
            BigInteger a;
            do
            {
                rng.GetBytes(bytes);
                a = new BigInteger(bytes, isUnsigned: true);
            }
            while (a < 2 || a >= n - 2);

            BigInteger x = BigInteger.ModPow(a, d, n);
            if (x == 1 || x == n - 1)
                continue;

            bool passedLoop = false;
            for (int j = 0; j < s - 1; j++)
            {
                x = BigInteger.ModPow(x, 2, n);
                if (x == 1) return false;
                if (x == n - 1)
                {
                    passedLoop = true;
                    break;
                }
            }

            if (!passedLoop)
                return false;
        }

        return true;
    }
}